pydis_core/exts/source.py

```diff
@@ -0,0 +1,198 @@
+            return
+
+        obj, source_type = await self._get_source_object(ctx, source_item)
+        embed = await self._build_embed(obj, source_type)
+        await ctx.send(embed=embed)
+
+    @staticmethod
+    async def _get_source_object(ctx: commands.Context["Bot"], argument: str) -> tuple[object, _SourceType]:
+        """Convert argument into the source object and source type."""
+        if argument.lower() == "help":
+            return ctx.bot.help_command, _SourceType.help_command
+
+        cog = ctx.bot.get_cog(argument)
+        if cog:
+            if cog.__module__.startswith("pydis_core.exts"):
+                return cog, _SourceType.core_cog
+            return cog, _SourceType.cog
+
+        cmd = ctx.bot.get_command(argument)
+        if cmd:
+            if cmd.module.startswith("pydis_core.exts"):
+                return cmd, _SourceType.core_command
+            return cmd, _SourceType.text_command
+
+        tags_cog = ctx.bot.get_cog("Tags")
+        show_tag = True
+
+        if not tags_cog:
+            show_tag = False
+        else:
+            identifier = _TagIdentifierStub.from_string(argument.lower())
+            if identifier in tags_cog.tags:
+                return identifier, _SourceType.tag
+
+        escaped_arg = escape_markdown(argument)
+
+        raise commands.BadArgument(
+            f"Unable to convert '{escaped_arg}' to valid command{', tag,' if show_tag else ''} or Cog."
+        )
+
+    def _get_source_link(self, source_item: object, source_type: _SourceType) -> tuple[str, str, int | None]:
+        """
+        Build GitHub link of source item, return this link, file location and first line number.
+
+        Raise BadArgument if `source_item` is a dynamically-created object (e.g. via internal eval).
+        """
+        if source_type == _SourceType.text_command or source_type == _SourceType.core_command:
+            source_item = inspect.unwrap(source_item.callback)
+            src = source_item.__code__
+            filename = src.co_filename
+        elif source_type == _SourceType.tag:
+            tags_cog = self.bot.get_cog("Tags")
+            filename = tags_cog.tags[source_item].file_path
+        else:
+            src = type(source_item)
+            try:
+                filename = inspect.getsourcefile(src)
+            except TypeError:
+                raise commands.BadArgument("Cannot get source for a dynamically-created object.")
+
+        if source_type != _SourceType.tag:
+            try:
+                lines, first_line_no = inspect.getsourcelines(src)
+            except OSError:
+                raise commands.BadArgument("Cannot get source for a dynamically-created object.")
+
+            lines_extension = f"#L{first_line_no}-L{first_line_no+len(lines)-1}"
+        else:
+            first_line_no = None
+            lines_extension = ""
+
+        # Handle tag file location differently than others to avoid errors in some cases
```

Comment:
outdated comment

Reply:
It is still somewhat relevant, just not here.

Reply:
304fe53b3200641905b18ca8cba52a5416f36503